File: app/src/main/java/com/example/myscope/activities/SignUp_Page.java
package com.example.myscope.activities;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;

import com.example.myscope.R;


public class SignUp_Page extends BaseActivity implements View.OnClickListener {

    private static final String TAG = "SignUp_Page";

    Button btn_SignUp;
    TextView btn_login;
    EditText edt_firstname, edt_lastname, edt_mobile, edt_email;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.signuppage_main);

        btn_SignUp = (Button) findViewById(R.id.btn_SignUp);
        btn_login = (TextView) findViewById(R.id.btn_login);


        edt_firstname = (EditText) findViewById(R.id.edt_firstname);
        edt_lastname = (EditText) findViewById(R.id.edt_lastname);
        edt_mobile = (EditText) findViewById(R.id.edit_mobile);
        edt_email = (EditText) findViewById(R.id.edt_email);

        btn_SignUp.setOnClickListener(this);
        btn_login.setOnClickListener(this);

    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {

            case R.id.btn_SignUp:
                signup();

                break;
            case R.id.btn_login:

                Intent login_btn = new Intent(getApplicationContext(), Login_Page.class);
                login_btn.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                startActivity(login_btn);

                break;
        }
    }

    private void signup() {
        Log.d(TAG, "SignUp_Page");

        String firstName = edt_firstname.getText().toString();
        String lastName = edt_lastname.getText().toString();
        String mobileNumber = edt_mobile.getText().toString();
        String emailId = edt_email.getText().toString();

        if (validate() == false) {

            onSignupFailed();
            return;
        }
//        sendPost(name, email, password, mobile);
        Intent login_btn = new Intent(getApplicationContext(), Login_Page.class);
        login_btn.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(login_btn);
        Toast.makeText(getApplicationContext(), "Registration Successfully Completed", Toast.LENGTH_LONG).show();
    }

    private void onSignupFailed() {

        Toast.makeText(getBaseContext(), "Please Enter all fields", Toast.LENGTH_LONG).show();

        btn_SignUp.setEnabled(true);

    }

    private boolean validate() {


        boolean valid = true;

        String firstName = edt_firstname.getText().toString();
        String lastName = edt_lastname.getText().toString();
        String mobileNumber = edt_mobile.getText().toString();
        String emailId = edt_email.getText().toString();

//        String TakenUsername = api.toString();

        if (firstName.isEmpty() || firstName.length() < 3) {
            edt_firstname.setError("at least 3 characters");
            valid = false;
        } else {
            edt_firstname.setError(null);
        }

        if (lastName.isEmpty() || lastName.length() < 3) {
            edt_lastname.setError("at least 3 characters");
            valid = false;
        } else {
            edt_lastname.setError(null);
        }


        if (mobileNumber.isEmpty() || mobileNumber.length() < 10) {
            edt_mobile.setError("enter mobile number");
            valid = false;
        } else {
            edt_mobile.setError(null);
        }

        if (emailId.isEmpty() || !android.util.Patterns.EMAIL_ADDRESS.matcher(emailId).matches()) {
            edt_email.setError("enter a valid email address");
            valid = false;
        } else {
            edt_email.setError(null);
        }

        return valid;
    }


}


//    private void sendPost(String name, String email, String password, String mobile) {
//
//        api.registeruser(email, password, name, mobile)
//                .enqueue(new Callback<LoginResponse>() {
//                    @Override
//                    public void onResponse(Call<LoginResponse> call, Response<LoginResponse> response) {
//
//                        Toast.makeText(getApplicationContext(), "Registration Success fully completed! Please Login", Toast.LENGTH_LONG).show();
//                        Intent intent = new Intent(getApplicationContext(), Login_Page.class);
//                        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
//                        startActivity(intent);
//                        Log.i(TAG, "post submitted to API." + response.body().toString());
//
//                    }
//
//                    @Override
//                    public void onFailure(Call<LoginResponse> call, Throwable t) {
//
//                        Toast.makeText(getApplicationContext(), "Please Enter Valid Login Details", Toast.LENGTH_LONG).show();
////                        Log.e(TAG, "Unable to submit post to API.");
//                    }
//                });
//
//    }

File: app/src/main/java/com/example/myscope/activities/SignUp_Page.kt
package com.example.myscope.activities

import android.app.Dialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.util.Patterns
import android.view.View
import android.view.Window
import android.widget.CheckBox
import com.example.myscope.R
import com.example.myscope.services.PrescriptionInterface
import com.google.gson.JsonParser
import kotlinx.android.synthetic.main.signuppage_main.*
import kotlinx.android.synthetic.main.terms_conditions_main.*
import org.json.JSONException
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


class SignUp_Page : BaseActivity(), View.OnClickListener {
    var firstName: String = "";
    var lastName: String = "";
    var mobileNumber: String = "";
    var emailId: String = "";
    var terms_conditions: CheckBox? = null
    private var registerCall: Call<SignupResponse>? = null
    private var request: PrescriptionInterface? = null

    lateinit var jsonParser: JsonParser
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.signuppage_main)
        terms_conditions = findViewById(R.id.terms_conditions)
        btn_SignUp!!.setOnClickListener(this)
        btn_login!!.setOnClickListener(this)
        terms_conditions_text!!.setOnClickListener(this)
        hideKeyBoard()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_SignUp -> signup()
            R.id.btn_login -> {
                navigateToActivity(Intent(applicationContext, Login_Page::class.java))
            }
            R.id.terms_conditions_text -> {
                showDialog()
            }
        }
    }

    private fun showDialog() {

        val d = Dialog(this)
        //NO TITLE
        d.requestWindowFeature(Window.FEATURE_NO_TITLE)
        //layout of dialog
        d.setContentView(R.layout.terms_conditions_main)
        d.ok_button.setOnClickListener{

            val intent = intent
            finish()
            startActivity(intent)
        }

        d.show()
    }

    private fun signup() {
        Log.d(TAG, "SignUp_Page")


        if (validate().equals(true)) {
            registerRetrofit2Api(firstName, lastName, mobileNumber, emailId);

        } else {

            onSignupFailed()
            return
        }

    }

    private fun onSignupFailed() {
        btn_SignUp!!.isEnabled = true
    }


    private fun validate(): Boolean {
        var valid = true

        firstName = edt_firstname!!.text.toString().trim { it <= ' ' }
        lastName = edt_lastname!!.text.toString().trim { it <= ' ' }
        mobileNumber = edt_mobile!!.text.toString().trim { it <= ' ' }
        emailId = edt_email!!.text.toString().trim { it <= ' ' }

        firstname_layout!!.error = null
        lastname_layout!!.error = null
        mobile_layout!!.error = null
        email_layout!!.error = null


        if (firstName.isEmpty()) {
            firstname_layout!!.error = "Please Enter First Name"
            valid = false

        } else if (firstName.length < 3) {
            firstname_layout!!.error = "At least 3 characters"
            valid = false
        }
        if (lastName.isEmpty()) {
            lastname_layout!!.error = "Please Enter Last Name"
            valid = false
        } else if (lastName.length < 3) {
            lastname_layout!!.error = "At least 3 characters"
            valid = false
        }
        if (mobileNumber.isEmpty()) {
            mobile_layout!!.error = "Enter mobile number"
            valid = false
        } else if (mobileNumber.length < 10) {
            mobile_layout!!.error = "Enter Valid mobile number"
            valid = false
        }
        if (emailId.isEmpty() ) {
            email_layout!!.error = "Enter Register Email Address"
            valid = false
        } else if(!Patterns.EMAIL_ADDRESS.matcher(emailId).matches()){
            email_layout!!.error = "Enter correct Email Address"
            valid = false
        }

        if (!terms_conditions!!.isChecked()) {
            valid = false
            showLongSnackBar("Please agree Terms&Condtions and Continue")
        }
        return valid
    }

    private fun registerRetrofit2Api(firstname: String, lastname: String, mobileno: String, email: String) {

        var login = SignupResponse(firstname, lastname, mobileno, email)
        request = APIClient.getClient().create(PrescriptionInterface::class.java)

        try {
            var resp: SignupResponse = login

            registerCall = request?.createPatient(resp)

            (registerCall as Call<SignupResponse>).enqueue(object : Callback<SignupResponse> {
                override fun onFailure(call: Call<SignupResponse>, t: Throwable) {
                    showLongToast("Recheck and save again")
                    Log.d("Errormessage", t.message + "error")
                }

                override fun onResponse(call: Call<SignupResponse>, resp: Response<SignupResponse>) {
                    if (resp.isSuccessful && !resp.equals("")) {
                        showShortToast("Api call is working")
                        navigateToActivity(Intent(applicationContext, Login_Page::class.java))
                    }
                }
            });
        } catch (e: JSONException) {
            e.printStackTrace();
        }
    }

    companion object {
        private const val TAG = "SignUp_Page"
    }


}
